esx_service/vmdk_ops.py

```diff
@@ -1703,6 +1704,101 @@ def wait_for_tasks(si, tasks):
+    # Add the property specs
+    propSpec = vmodl.query.PropertyCollector.PropertySpec(type=vim.VirtualMachine, all=False)
+    propSpec.pathSet.append('runtime.powerState')
+    filterSpec.propSet.append(propSpec)
+    try:
+        pcFilter = pc.CreateFilter(filterSpec, True)
+        atexit.register(pcFilter.Destroy)
+    except Exception as e:
+        logging.error("Problem creating PropertyCollector %s", str(e))
+
+def listen_vm_propertychange(pc, empty):
+    logging.info("PropertyChangeListener thread started")
+    version = ''
+    while True:
+        result = pc.WaitForUpdates(version)
+
+        # if the wait timed out, wait again
+        if result is None:
+            continue
+
+        try:
+            # process the updates result
+            for filterSet in result.filterSet:
+                for objectSet in filterSet.objectSet:
+                    moref = getattr(objectSet, 'obj', None)
+                    if objectSet.kind == 'modify':
+                        for change in objectSet.changeSet:
+                            if change.name == 'runtime.powerState' and change.val == 'poweredOff':
+                                vm_name = moref.config.name
+                                logging.info("VM poweroff change found for %s", vm_name)
+                                # if the event was powerOff for a VM, set the status of all
+                                # docker volumes attached to the VM to be detached
+                                set_device_detached(moref.config.hardware.device)
+        except Exception as e:
+            logging.error("PropertyChangeListener: error %s", str(e))
+
+        version = result.version
+
+    logging.info("PropertyChangeListener thread exiting")
+
+def vm_folder_traversal():
+    """
+    Build the traversal spec for the property collector to traverse vmFolder
+    """
+
+    TraversalSpec = vmodl.query.PropertyCollector.TraversalSpec
+    SelectionSpec = vmodl.query.PropertyCollector.SelectionSpec
+
+    # Traversal through vmFolder branch
+    dcToVmf = TraversalSpec(name='dcToVmf', type=vim.Datacenter, path='vmFolder', skip=False)
+    dcToVmf.selectSet.append(SelectionSpec(name='visitFolders'))
+
+    # Recurse through the folders
+    visitFolders = TraversalSpec(name='visitFolders', type=vim.Folder,path='childEntity', skip=False)
+    visitFolders.selectSet.extend((SelectionSpec(name='visitFolders'), SelectionSpec(name='dcToVmf'),))
+
+    return SelectionSpec.Array((visitFolders, dcToVmf,))
+
+def set_device_detached(device_list):
+    """
+    For all devices in device_list, if it is a docker volume, set its status to detached in KV
+    """
+
+    if not device_list:
+        return
+
+    try:
+        for dev in device_list:
+            # if it is a docker volume, construct the vmdk_path and set its status as detached
+            if vmdk_utils.check_docker_volume(dev):
+                datastore, disk_path = dev.backing.fileName.rsplit("]", 1)
```

Comment:
Pls. add a comment here showing the format of "dev" and the diskpath and datastore formats. Just so its clear how these are split/joined.